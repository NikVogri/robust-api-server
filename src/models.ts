import { Request as ExpressRequest, Response as ExpressResponse } from 'express';

export interface Request<Param = object, Body = object, Query = object>
    extends ExpressRequest<Param, object, Body, Query> {}

export interface Response<Res = object, Locals extends Record<string, unknown> = Record<string, unknown>>
    extends ExpressResponse<Res, Locals> {}

export interface CatDBItem {
    id: number;
    name: string;
    date_of_birth: string;
    color: string;
    created_at: string;
    updated_at: string;
}
