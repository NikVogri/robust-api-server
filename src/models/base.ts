import { Postgres } from '../clients/postgres';

export abstract class Model<T> {
    constructor(
        protected table: string,
        protected db: Postgres,
    ) {}

    async getAll(): Promise<T[]> {
        const { rows } = await this.db.pool.query(`SELECT * FROM ${this.table}`);
        return rows;
    }

    async getById(id: string): Promise<T> {
        const { rows } = await this.db.pool.query(`SELECT * FROM ${this.table} WHERE id = $1`, [id]);
        return rows[0];
    }

    async delete(id: string): Promise<void> {
        await this.db.pool.query(`DELETE FROM ${this.table} WHERE id = $2`, [id]);
    }
}
