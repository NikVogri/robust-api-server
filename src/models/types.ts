export interface Cat {
    id: number;
    name: string;
    date_of_birth: string;
    color: string;
    created_at: string;
    updated_at: string;
}
