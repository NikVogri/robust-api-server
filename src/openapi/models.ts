/* eslint-disable */
/* tslint:disable */
/*
 * ---------------------------------------------------------------
 * ## THIS FILE WAS GENERATED VIA SWAGGER-TYPESCRIPT-API        ##
 * ##                                                           ##
 * ## AUTHOR: acacode                                           ##
 * ## SOURCE: https://github.com/acacode/swagger-typescript-api ##
 * ---------------------------------------------------------------
 */

export interface Error {
    message?: string;
}

export interface Cat {
    name: string;
    date_of_birth: string;
    color: string;
    created_at: string;
    updated_at: string;
}
