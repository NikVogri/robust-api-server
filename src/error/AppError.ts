interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
