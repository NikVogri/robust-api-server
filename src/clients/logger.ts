import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
