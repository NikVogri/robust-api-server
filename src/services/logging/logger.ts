import winston from 'winston';
import { AppError } from '../../error/AppError';
import { singleton } from 'tsyringe';
import { Request, Response } from '../../types';

@singleton()
export class Logger {
    public readonly winston: winston.Logger;

    private readonly custom = {
        levels: {
            error: 0,
            info: 1,
            http: 2,
            verbose: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
            verbose: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.winston = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'verbose',
                format: winston.format.combine(
                    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                    winston.format.printf(
                        (i) => `${i.level} ${i.timestamp}: ${i.message} ${JSON.stringify(i.stack) ?? ''}`,
                    ),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error) {
        this.winston.error(message, AppError.toStringifiableObject(error));
    }

    info(message: string) {
        this.winston.info(message);
    }

    http(req: Request, res: Response, error?: Error) {
        const log = this.buildHttpLog(req, res, error);
        this.winston.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: unknown): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: Object.keys(req.query).length > 0 ? req.query : undefined,
                params: Object.keys(req.params).length > 0 ? req.params : undefined,
                body: Object.keys(req.body).length > 0 ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: this.buildErrorObject(error),
        });
    }

    private buildErrorObject(error?: unknown) {
        if (!error || !(error instanceof Error)) return undefined;

        const output: Record<string, unknown> = {
            message: error?.message,
        };

        if (error instanceof AppError && error.thrownError) {
            output['thrown'] = AppError.toStringifiableObject(error.thrownError);
        }

        return output;
    }
}
