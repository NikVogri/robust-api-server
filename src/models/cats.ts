import { Model } from './base';
import { Cat } from './types';
import { Postgres } from '../clients/postgres';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';

@injectable()
export class CatsModel extends Model<Cat> {
    constructor(db: Postgres) {
        super('cats', db);
    }

    async create(createCatDto: CreateCatDto): Promise<Cat> {
        const { rows } = await this.db.pool.query(
            `
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                `,
            [createCatDto.name, createCatDto.dateOfBirth, createCatDto.color],
        );

        return rows[0];
    }
}
