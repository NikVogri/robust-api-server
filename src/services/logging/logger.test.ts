import { describe, expect, test, vi } from 'vitest';
import { Logger } from './logger';
import winston from 'winston';
import { Request, Response } from '../../models';
import { AppError } from '../../error/AppError';

describe('Test Logger client', () => {
    const logger = new Logger();

    test('error logging', async () => {
        vi.spyOn(logger.winston, 'error').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return {} as winston.Logger;
        });

        const error = new Error('Something went wrong');
        error.stack = 'error-stack';

        const logMessage = 'Oh, uh!';

        logger.error(logMessage, error);
    });

    test('info logging', async () => {
        vi.spyOn(logger.winston, 'info').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return {} as winston.Logger;
        });

        logger.info('This is some info');
    });

    test('http logging', async () => {
        vi.spyOn(logger.winston, 'http').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return {} as winston.Logger;
        });

        const request = {
            originalUrl: '/some/path/:id/items',
            url: '/some/path/:id/items',
            path: '/some/path/:id/items',
            method: 'GET',
            hostname: 'hostname',
            version: 'HTTP/1.1',
            headers: {
                ['user-agent']: 'test',
            },
            query: {
                page: '1',
                perPage: '10',
            },
            params: {
                id: '12345',
            },
            body: {
                session: '987654321',
            },
        } as unknown as Request;

        const response = {
            statusCode: 400,
        } as Response;

        const thrownError = new Error('Something went wrong');
        thrownError.stack = 'thrown-error-stack';

        const appError = new AppError({
            message: 'this is a message',
            statusCode: 400,
            thrownError: thrownError,
        });

        logger.http(request, response, appError);
    });
});
