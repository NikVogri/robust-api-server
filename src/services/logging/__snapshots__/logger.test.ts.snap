// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Logger client > error logging 1`] = `
[
  "Oh, uh!",
  {
    "message": "Something went wrong",
    "stack": "error-stack",
  },
]
`;

exports[`Test Logger client > http logging 1`] = `
[
  "{"success":false,"originalUrl":"/some/path/:id/items","path":"GET /some/path/:id/items","client":{"hostname":"hostname","version":"HTTP/1.1","userAgent":"test"},"req":{"query":{"page":"1","perPage":"10"},"params":{"id":"12345"},"body":["session"]},"res":{"statusCode":400},"error":{"message":"this is a message","thrownError":{"stack":"thrown-error-stack","message":"Something went wrong"}}}",
]
`;

exports[`Test Logger client > info logging 1`] = `
[
  "This is some info",
]
`;
